src/core/include/mp-units/bits/fixed_point.h

```diff
@@ -0,0 +1,228 @@
+  static constexpr std::size_t base_width = std::numeric_limits<std::make_unsigned_t<T>>::digits;
+  static constexpr std::size_t width = 2 * base_width;
+
+  using Th = T;
+  using Tl = std::make_unsigned_t<T>;
+
+  constexpr double_width_int() = default;
+  constexpr double_width_int(const double_width_int&) = default;
+
+  constexpr double_width_int& operator=(const double_width_int&) = default;
+
+  constexpr double_width_int(Th hi_, Tl lo_) : hi(hi_), lo(lo_) {}
+
+  explicit(true) constexpr double_width_int(long double v)
+  {
+    constexpr auto scale = int_power<long double>(2, base_width);
+    constexpr auto iscale = 1.l / scale;
+    hi = static_cast<Th>(v * iscale);
+    lo = static_cast<Tl>(v - (hi * scale));
+  }
+  template<std::integral U>
+    requires(is_signed || !std::is_signed_v<U>)
+  explicit(false) constexpr double_width_int(U v)
+  {
+    if constexpr (is_signed) {
+      hi = v < 0 ? Th{-1} : Th{0};
+    } else {
+      hi = 0;
+    }
+    lo = static_cast<Tl>(v);
+  }
+
+  explicit(true) constexpr operator Th() const { return static_cast<Th>(lo); }
+
+  constexpr auto operator<=>(const double_width_int&) const = default;
+
+  // calculates the double-width product of two base-size integers
+  static constexpr double_width_int wide_product_of(Th lhs, Tl rhs)
+  {
+    constexpr std::size_t half_width = base_width / 2;
+    constexpr Tl msk = Tl(1) << half_width;
+    Th l1 = lhs >> half_width;
+    Tl l0 = static_cast<Tl>(lhs) & msk;
+    Tl r1 = rhs >> half_width;
+    Tl r0 = rhs & msk;
+    Tl t00 = l0 * r0;
+    Tl t01 = l0 * r1;
+    Th t10 = l1 * r0;
+    Th t11 = l1 * r1;
+    Tl m = (t01 & msk) + (static_cast<Tl>(t10) & msk) + (t00 >> half_width);
+    Th o1 = t11 + (m >> half_width) + (t10 >> half_width) + static_cast<Th>(t01 >> half_width);
+    Tl o0 = (t00 & msk) | ((m & msk) << half_width);
+    return {o1, o0};
+  }
+
+  template<std::integral Lhs>
+  friend constexpr auto operator*(Lhs lhs, const double_width_int& rhs)
+  {
+    using ret_t = double_width_int<std::common_type_t<Lhs, Th>>;
+    auto ret = ret_t::wide_product_of(lhs, rhs.lo);
+    ret.hi += lhs * rhs.hi;
+    return ret;
+  };
+
+
+  constexpr double_width_int operator+(Tl rhs) const
+  {
+    auto ret = lo + rhs;
+    return {hi + (ret < lo ? 1 : 0), ret};
+  }
+
+  constexpr double_width_int operator>>(unsigned n) const
+  {
+    if (n >= base_width) {
+      return {0, hi >> (n - base_width)};
+    }
+    return {hi >> n, (static_cast<Tl>(hi) << (base_width - n)) | (lo >> n)};
+  }
+  constexpr double_width_int operator<<(unsigned n) const
+  {
+    if (n >= base_width) {
+      return {static_cast<Th>(lo >> (2 * base_width - n)), 0};
+    }
+    return {(hi << n) + static_cast<Th>(lo >> (base_width - n)), lo << n};
+  }
+
+  static constexpr double_width_int max() { return {std::numeric_limits<Th>::max(), std::numeric_limits<Tl>::max()}; }
+
+  Th hi;
+  Tl lo;
+};
+
+#if false && defined(__SIZEOF_INT128__)
```

Comment:
yeah, CI complains about the true-path because `__int128_t` is non-standard. Given that this is an implementation-detail, I should probably look for a way to disable that warning instead.